Fix Observation Download

JoeCohen opened the pull request:
- Allows Observation Download if the number of Observatios downloaded is reasonable, or if in admin_mode.
- Delivers #2553

JoeCohen committed:
Fix Observation Download

- Allows Observation Download if the number of Observatios downloaded is reasonable, or if in admin_mode

JoeCohen commented:
Questions:
- What's the right limit on # of Observations to download?
- @pellaea suggests stubbing `MAX_DOWNLOADS` instead of creating that many Obss in the tests.  See #2553
What's best?

pellaea commented on app/controllers/observations/downloads_controller.rb:
```diff
@@ -4,6 +4,8 @@ module Observations
   class DownloadsController < ApplicationController
     before_action :login_required
 
+    MAX_DOWNLOADS = 120
```
120 seems way too low.  Surely there's some compromise between half a million and 120?  It was originally designed to be able to render a million in a few seconds, so surely 1000 wouldn't be too many?  Has anyone actually tested the performance of this method?

JoeCohen commented:
Good question.
1. I'm not aware of performance tests.
BUT
2. Downloading all Observations was bringing the site to its knees.

On Wed, Dec 4, 2024 at 8:20 PM Jason Hollinger ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In app/controllers/observations/downloads_controller.rb
> <https://github.com/MushroomObserver/mushroom-observer/pull/2561#discussion_r1870597301>
> :
>
> > @@ -4,6 +4,8 @@ module Observations
>    class DownloadsController < ApplicationController
>      before_action :login_required
>
> +    MAX_DOWNLOADS = 120
>
> 120 seems way too low. Surely there's some compromise between half a
> million and 120? It was originally designed to be able to render a million
> in a few seconds, so surely 1000 wouldn't be too many? Has anyone actually
> tested the performance of this method?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/2561#pullrequestreview-2480386522>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFENGQOTWR7RUADG7GL2D7IAZAVCNFSM6AAAAABTBSA2VKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDIOBQGM4DMNJSGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

JoeCohen commented:
I'll bump the number. Let me give it some thought. I'll probably do it
tomorrow.

On Wed, Dec 4, 2024 at 8:23 PM Joseph Cohen ***@***.***> wrote:

> Good question.
> 1. I'm not aware of performance tests.
> BUT
> 2. Downloading all Observations was bringing the site to its knees.
>
> On Wed, Dec 4, 2024 at 8:20 PM Jason Hollinger ***@***.***>
> wrote:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In app/controllers/observations/downloads_controller.rb
>> <https://github.com/MushroomObserver/mushroom-observer/pull/2561#discussion_r1870597301>
>> :
>>
>> > @@ -4,6 +4,8 @@ module Observations
>>    class DownloadsController < ApplicationController
>>      before_action :login_required
>>
>> +    MAX_DOWNLOADS = 120
>>
>> 120 seems way too low. Surely there's some compromise between half a
>> million and 120? It was originally designed to be able to render a million
>> in a few seconds, so surely 1000 wouldn't be too many? Has anyone actually
>> tested the performance of this method?
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/MushroomObserver/mushroom-observer/pull/2561#pullrequestreview-2480386522>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAALDFENGQOTWR7RUADG7GL2D7IAZAVCNFSM6AAAAABTBSA2VKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDIOBQGM4DMNJSGI>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***
>> com>
>>
>

pellaea commented:
Fair enough. Maybe start with 100, push it out, test it a few times and
look at the log to see how long 100 takes.  Then bump it to 1000.  Rinse
and repeat until it pushes against what you think is acceptable.

Long term, it probably makes sense to convert it into an async job, right?
Probably pretty easy to do, at this point, too.  We've got a number of
examples of async jobs -- field slips, inat import, original image
retrieval.  Probably just need some sort of job tracker table in the
database and move a few files and tests around.

On Wed, Dec 4, 2024 at 11:26 PM Joseph D. Cohen ***@***.***>
wrote:

> I'll bump the number. Let me give it some thought. I'll probably do it
> tomorrow.
>
> On Wed, Dec 4, 2024 at 8:23 PM Joseph Cohen ***@***.***> wrote:
>
> > Good question.
> > 1. I'm not aware of performance tests.
> > BUT
> > 2. Downloading all Observations was bringing the site to its knees.
> >
> > On Wed, Dec 4, 2024 at 8:20 PM Jason Hollinger ***@***.***>
> > wrote:
> >
> >> ***@***.**** commented on this pull request.
> >> ------------------------------
> >>
> >> In app/controllers/observations/downloads_controller.rb
> >> <
> https://github.com/MushroomObserver/mushroom-observer/pull/2561#discussion_r1870597301>
>
> >> :
> >>
> >> > @@ -4,6 +4,8 @@ module Observations
> >> class DownloadsController < ApplicationController
> >> before_action :login_required
> >>
> >> + MAX_DOWNLOADS = 120
> >>
> >> 120 seems way too low. Surely there's some compromise between half a
> >> million and 120? It was originally designed to be able to render a
> million
> >> in a few seconds, so surely 1000 wouldn't be too many? Has anyone
> actually
> >> tested the performance of this method?
> >>
> >> —
> >> Reply to this email directly, view it on GitHub
> >> <
> https://github.com/MushroomObserver/mushroom-observer/pull/2561#pullrequestreview-2480386522>,
>
> >> or unsubscribe
> >> <
> https://github.com/notifications/unsubscribe-auth/AAALDFENGQOTWR7RUADG7GL2D7IAZAVCNFSM6AAAAABTBSA2VKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDIOBQGM4DMNJSGI>
>
> >> .
> >> You are receiving this because you authored the thread.Message ID:
> >> ***@***.***
> >> com>
> >>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/2561#issuecomment-2519065855>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYTNNORLRI36WTNDNMQJXT2D7IV5AVCNFSM6AAAAABTBSA2VKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJZGA3DKOBVGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

JoeCohen commented:
Good suggestions.
I agree that a job is the long-range solution.

On Wed, Dec 4, 2024 at 8:48 PM Jason Hollinger ***@***.***>
wrote:

> Fair enough. Maybe start with 100, push it out, test it a few times and
> look at the log to see how long 100 takes. Then bump it to 1000. Rinse
> and repeat until it pushes against what you think is acceptable.
>
> Long term, it probably makes sense to convert it into an async job, right?
> Probably pretty easy to do, at this point, too. We've got a number of
> examples of async jobs -- field slips, inat import, original image
> retrieval. Probably just need some sort of job tracker table in the
> database and move a few files and tests around.
>
> On Wed, Dec 4, 2024 at 11:26 PM Joseph D. Cohen ***@***.***>
> wrote:
>
> > I'll bump the number. Let me give it some thought. I'll probably do it
> > tomorrow.
> >
> > On Wed, Dec 4, 2024 at 8:23 PM Joseph Cohen ***@***.***> wrote:
> >
> > > Good question.
> > > 1. I'm not aware of performance tests.
> > > BUT
> > > 2. Downloading all Observations was bringing the site to its knees.
> > >
> > > On Wed, Dec 4, 2024 at 8:20 PM Jason Hollinger ***@***.***>
> > > wrote:
> > >
> > >> ***@***.**** commented on this pull request.
> > >> ------------------------------
> > >>
> > >> In app/controllers/observations/downloads_controller.rb
> > >> <
> >
> https://github.com/MushroomObserver/mushroom-observer/pull/2561#discussion_r1870597301>
>
> >
> > >> :
> > >>
> > >> > @@ -4,6 +4,8 @@ module Observations
> > >> class DownloadsController < ApplicationController
> > >> before_action :login_required
> > >>
> > >> + MAX_DOWNLOADS = 120
> > >>
> > >> 120 seems way too low. Surely there's some compromise between half a
> > >> million and 120? It was originally designed to be able to render a
> > million
> > >> in a few seconds, so surely 1000 wouldn't be too many? Has anyone
> > actually
> > >> tested the performance of this method?
> > >>
> > >> —
> > >> Reply to this email directly, view it on GitHub
> > >> <
> >
> https://github.com/MushroomObserver/mushroom-observer/pull/2561#pullrequestreview-2480386522>,
>
> >
> > >> or unsubscribe
> > >> <
> >
> https://github.com/notifications/unsubscribe-auth/AAALDFENGQOTWR7RUADG7GL2D7IAZAVCNFSM6AAAAABTBSA2VKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDIOBQGM4DMNJSGI>
>
> >
> > >> .
> > >> You are receiving this because you authored the thread.Message ID:
> > >> ***@***.***
> > >> com>
> > >>
> > >
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/MushroomObserver/mushroom-observer/pull/2561#issuecomment-2519065855>,
>
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAYTNNORLRI36WTNDNMQJXT2D7IV5AVCNFSM6AAAAABTBSA2VKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJZGA3DKOBVGU>
>
> > .
> > You are receiving this because you were mentioned.Message ID:
> > ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/2561#issuecomment-2519098504>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALDFGNKCTTBUGWJ4RSMDD2D7LIBAVCNFSM6AAAAABTBSA2VKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJZGA4TQNJQGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

JoeCohen committed:
Stub max_downloads instead of creating Obss

- Implements [@pellaea's suggestion](https://github.com/MushroomObserver/mushroom-observer/issues/2553#issuecomment-2519021969)
- Uses MO application variable `max_dowloads` an  instead of the constant `MAX_DOWNLOADS`
- Stubs the variable in the tests instead of creating many Observations
- Improves test speed. With that value set to 1_200, the relevant tests took 2 secs, versus 11 seconds if I created enough Observations to make the tests work.

JoeCohen committed:
DRY definition of max_downloads

- Moves it from from multiple enviroment files to `consts.rb`
